### PR TITLE
Absent optional headers become the empty string

### PR DESCRIPTION
**Pull request by markelliot**

### CODE REVIEW

**Comment by markelliot** [comment]
@pnepywoda @j-baker

**Comment by markelliot** [comment]
- @uschi2000: this fixes the incongruity you flagged, turns out not so bad

File: http-clients/src/main/java/com/palantir/remoting/http/GuavaEmptyOptionalExpander.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+package com.palantir.remoting.http;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import feign.Param.Expander;
+import java.util.Objects;
+
+/**
+ * Expands Optional by using null for Optional.absent and the toString() of the value otherwise.
```
Comment (pnepywoda) [inline http-clients/src/main/java/com/palantir/remoting/http/GuavaEmptyOptionalExpander.java]: "using empty string"

File: http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java [hunk 2]
```diff
@@ -114,7 +114,7 @@ public void testStringQuery() throws Exception {
     public void testAbsentHeader() throws Exception {
         proxy.header(Optional.<String>absent(), "str2");
         RecordedRequest takeRequest = server.takeRequest();
-        assertThat(takeRequest.getHeader("opt"), is("{opt}"));
```
Comment (bavardage) [inline http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java:117]: so we change behaviour here. Were things relying on the old behaviour?

Comment (markelliot) [inline http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java:117]: we just added this behavior ~an hour ago. so...no.

File: http-clients/src/main/java/com/palantir/remoting/http/GuavaEmptyOptionalExpander.java (hunk 1 above)
Comment (markelliot) [inline http-clients/src/main/java/com/palantir/remoting/http/GuavaEmptyOptionalExpander.java]: :white_check_mark:

File: http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java [hunk 3]
```diff
@@ -114,7 +114,7 @@ public void testStringQuery() throws Exception {
...
+        assertThat(takeRequest.getHeader("opt"), is(""));
```
Comment (pnepywoda) [inline http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java:117]: yeah I'd have liked to keep this out of the header params all-together but that requires feign changes

**Comment by pnepywoda** [comment]
+1

File: http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java (hunk 3 above)
Comment (markelliot) [inline http-clients/src/test/java/com/palantir/remoting/http/GuavaOptionalAwareContractTest.java:117]: I think it's desirable to expand Optional header params, no?

**Comment by uschi2000** [comment]
cheers

**Pull request merged**